src/renderer/base/CSSLengthPercentage.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+
+#include "../inc/CSSLengthPercentage.h"
+
+CSSLengthPercentage CSSLengthPercentage::FromString(const wchar_t* str)
+{
+    auto& errnoRef = errno; // Nonzero cost, pay it once.
+    errnoRef = 0;
+
+    wchar_t* end;
+    auto value = std::wcstof(str, &end);
+
+    if (str == end || errnoRef == ERANGE)
+    {
+        return {};
+    }
+
+    auto referenceFrame = ReferenceFrame::FontSize;
+
+    if (*end)
+    {
+        if (wcscmp(end, L"%") == 0)
+        {
+            value /= 100.0f;
+        }
+        else if (wcscmp(end, L"px") == 0)
+        {
+            referenceFrame = ReferenceFrame::Absolute;
+            value /= 96.0f;
+        }
+        else if (wcscmp(end, L"pt") == 0)
+        {
+            referenceFrame = ReferenceFrame::Absolute;
+            value /= 72.0f;
+        }
+        else if (wcscmp(end, L"ch") == 0)
+        {
+            referenceFrame = ReferenceFrame::AdvanceWidth;
+        }
+        else
+        {
+            return {};
+        }
+    }
+
+    CSSLengthPercentage obj;
+    obj._value = value;
+    obj._referenceFrame = referenceFrame;
+    return obj;
+}
+
+CSSLengthPercentage::ReferenceFrame CSSLengthPercentage::GetReferenceFrame() const noexcept
+{
+    return _referenceFrame;
+}
+
+float CSSLengthPercentage::Resolve(float factor) const noexcept
+{
+    return _value * factor;

```

Comment:
I don't love how we have two functions, same name, one of which uses the local reference frame and one of which doesn't. Gotta look more to see what it means